src/proof/retainer.rs

```diff
@@ -1,16 +1,83 @@
+#[derive(Default, Clone, Debug)]
+struct AddedRemovedKeysTracking {
+    /// Path of the last branch which was passed to `retain_branch_proof`, regardless of whether it
+    /// was a target or not.
+    last_branch_path: Nibbles,
+    /// Proof of the node at `last_branch_path`.
+    last_branch_proof: Vec<u8>,
+    /// Path of the last node which was not found in the `ProofRetainer`'s target set, and which is
+    /// not potentially removed as given by [`AddedRemovedKeys::get_removed_mask`]
+    last_nontarget_path: Nibbles,
+    /// Proof of the node at `last_nontarget_path`.
+    last_nontarget_proof: Vec<u8>,
+}
+
+impl AddedRemovedKeysTracking {
+    /// Stores the path and proof of the most recently seen path/proof which were not retained and
+    /// are not potentially removed.
+    fn track_nontarget(&mut self, path: &Nibbles, proof: &[u8]) {
+        trace!(target: "trie::proof_retainer", ?path, "Tracking non-target");
+        self.last_nontarget_path = *path;
+        self.last_nontarget_proof.clear();
+        self.last_nontarget_proof.extend(proof);
+    }
+
+    /// Stores the path and proof of the most recently seen branch.
+    fn track_branch(&mut self, path: &Nibbles, proof: &[u8]) {
+        trace!(target: "trie::proof_retainer", ?path, "Tracking branch");
+        self.last_branch_path = *path;
+        self.last_branch_proof.clear();
+        self.last_branch_proof.extend(proof);
+    }
+
+    /// Checks if an extension node is parent to the last tracked branch node.
+    fn extension_child_is_nontarget(&self, path: &Nibbles, short_key: &Nibbles) -> bool {
+        path.len() + short_key.len() == self.last_nontarget_path.len()
+            && self.last_nontarget_path.starts_with(path)
+            && self.last_nontarget_path.ends_with(short_key)
+    }
+
+    /// Checks if a branch node is parent to the last non-target node.
+    fn branch_child_is_nontarget(&self, path: &Nibbles, child_nibble: u8) -> bool {
+        path.len() + 1 == self.last_nontarget_path.len()
+            && self.last_nontarget_path.starts_with(path)
+            && self.last_nontarget_path.last().expect("path length >= 1") == child_nibble
+    }
+}
+
 /// Proof retainer is used to store proofs during merkle trie construction.
 /// It is intended to be used within the [`HashBuilder`](crate::HashBuilder).
+///
+/// When using the `retain_leaf_proof`, `retain_extension_proof`, and `retain_branch_proof`
+/// methods, it is required that the calls are ordered such that proofs of parent nodes are
+/// retained after their children.
 #[derive(Default, Clone, Debug)]
-pub struct ProofRetainer {
+pub struct ProofRetainer<K = AddedRemovedKeys> {
     /// The nibbles of the target trie keys to retain proofs for.
     targets: Vec<Nibbles>,
     /// The map retained trie node keys to RLP serialized trie nodes.
     proof_nodes: ProofNodes,
+    /// Provided by the user to give the necessary context to retain extra proofs.
+    added_removed_keys: Option<K>,
```

Comment:
why do we have a generic here? I think we could just have this be `AddedRemovedKeys`, although if it needs to be shared maybe it should be a reference type?